src/sampler.jl

```diff
@@ -0,0 +1,60 @@
+#=
+    `MVNSampler` is used to draw from multivariate normal distribution
+=#
+
+import Base: ==
+
+immutable MVNSampler{TM<:Real,TS<:Real,TQ<:LinAlg.BlasReal}
+    mu::Vector{TM}
+    Sigma::Matrix{TS}
+    Q::Matrix{TQ}
+end
+
+function MVNSampler{TM<:Real,TS<:Real}(mu::Vector{TM}, Sigma::Matrix{TS})
+    ATOL1, RTOL1 = 1e-8, 1e-8
+    ATOL2, RTOL2 = 1e-8, 1e-14
+
+    n = length(mu)
+
+    if size(Sigma) != (n,n) # Check Sigma is n x n
+        throw(ArgumentError("Sigma must be 2 dimensional and square matrix of same length to mu"))
+    end
+
+    issymmetric(Sigma) || throw(ArgumentError("Sigma must be symmetric"))
+
+    C = cholfact(Symmetric(Sigma, :L), Val{true})
+
+    if C.rank == n  # Positive definite
+        p = invperm(C.piv)
+        Q = tril(C.factors)[p,p]
+        return MVNSampler(mu, Sigma, Q)
+    end
+
+    non_PSD_msg = "Sigma must be positive semidefinite"
+
+    for i in C.rank+1:n
+        if C[:L][i, i] < -ATOL1 || C[:L][i, i]/norm(diag(C[:L])) < -RTOL1 # Not positive semidefinite
+            throw(ArgumentError(non_PSD_msg))
+        end
+    end
+
+    p = invperm(C.piv)
+    Q = tril(C.factors)[p,p]
+
+    if maxabs(Sigma - Q * Q') > ATOL2 && maxabs(Sigma - Q * Q')/max(norm(Sigma),norm(Q * Q')) > RTOL2 # Not positive semidefinite

```

Comment:
1. If we want to exactly follow the [NumPy code on `master`](https://github.com/numpy/numpy/blob/master/numpy/random/mtrand/mtrand.pyx#L4506) (which is more recent than the code in the link provided by @Shunsuke-Hori), it will be:
   
   ```jl
   Q_2 = Q * Q'
   for j in 1:n
       for i in 1:n
           abs(Q_2[i, j] - Sigma[i, j]) <= ATOL2 + RTOL2 * abs(Sigma[i, j]) ||
               throw(ArgumentError(non_PSD_msg))
       end
   end
   ```

   where `ATOL2` and `RTOL2` default to `1e-8`.

2. Alternatively, we can just use [`isapprox`](https://docs.julialang.org/en/stable/stdlib/math/#Base.isapprox):
   
   ```jl
   isapprox(Q*Q', Sigma; rtol=RTOL2, atol=ATOL2) ||
       throw(ArgumentError(non_PSD_msg))
   ```